HPXMLtoOpenStudio/resources/hvac_sizing.rb

```diff
@@ -2685,325 +2672,99 @@ def self.gshp_hxbore_ft_per_ton(weather, hvac_cooling_ap, bore_spacing, pipe_r_v
-      if num_bore_holes == 8
-        if spacing_to_depth_ratio <= 0.02
-          gfnc_coeff = 2.684, 3.066, 3.497, 4.275, 5.229, 6.767, 8.724, 10.417, 12.723, 14.079, 15.03, 16.332, 17.217, 19.17, 20.835, 21.288, 21.698, 22.101, 22.417, 22.692, 22.931, 23.133, 23.286, 23.462, 23.583, 23.642, 23.742
-        elsif spacing_to_depth_ratio <= 0.03
-          gfnc_coeff = 2.679, 3.026, 3.347, 3.821, 4.409, 5.418, 6.87, 8.226, 10.299, 11.565, 12.466, 13.716, 14.58, 16.498, 18.125, 18.572, 18.972, 19.368, 19.679, 19.946, 20.179, 20.376, 20.527, 20.699, 20.816, 20.874, 20.967
-        elsif spacing_to_depth_ratio <= 0.05
-          gfnc_coeff = 2.679, 3.023, 3.319, 3.673, 4.018, 4.564, 5.389, 6.21, 7.763, 8.801, 9.582, 10.709, 11.51, 13.311, 14.912, 15.349, 15.744, 16.13, 16.432, 16.693, 16.921, 17.114, 17.259, 17.426, 17.54, 17.595, 17.686
-        elsif spacing_to_depth_ratio <= 0.1
-          gfnc_coeff = 2.679, 3.023, 3.318, 3.664, 3.962, 4.313, 4.704, 4.999, 5.725, 6.294, 6.771, 7.543, 8.14, 9.629, 11.105, 11.52, 11.908, 12.28, 12.578, 12.831, 13.054, 13.244, 13.386, 13.548, 13.659, 13.712, 13.8
-        else
-          gfnc_coeff = 2.679, 3.023, 3.318, 3.664, 3.961, 4.306, 4.651, 4.839, 5.293, 5.641, 5.938, 6.44, 6.856, 8.062, 9.297, 9.681, 10.036, 10.394, 10.692, 10.941, 11.163, 11.35, 11.492, 11.654, 11.766, 11.819, 11.907
-        end
-      elsif num_bore_holes == 10
-        if spacing_to_depth_ratio <= 0.02
-          gfnc_coeff = 2.684, 3.066, 3.494, 4.262, 5.213, 6.81, 8.965, 10.906, 13.643, 15.283, 16.443, 18.038, 19.126, 21.532, 23.581, 24.138, 24.642, 25.137, 25.525, 25.862, 26.155, 26.403, 26.59, 26.806, 26.955, 27.027, 27.149
-        elsif spacing_to_depth_ratio <= 0.03
-          gfnc_coeff = 2.679, 3.026, 3.346, 3.818, 4.399, 5.4, 6.889, 8.358, 10.713, 12.198, 13.27, 14.776, 15.824, 18.167, 20.158, 20.704, 21.194, 21.677, 22.057, 22.382, 22.666, 22.907, 23.09, 23.3, 23.443, 23.513, 23.627
-        elsif spacing_to_depth_ratio <= 0.05
-          gfnc_coeff = 2.679, 3.023, 3.319, 3.673, 4.018, 4.559, 5.374, 6.193, 7.814, 8.951, 9.831, 11.13, 12.069, 14.219, 16.154, 16.684, 17.164, 17.631, 17.998, 18.314, 18.59, 18.824, 19, 19.201, 19.338, 19.405, 19.515
-        elsif spacing_to_depth_ratio <= 0.1
-          gfnc_coeff = 2.679, 3.023, 3.318, 3.664, 3.962, 4.313, 4.703, 4.996, 5.712, 6.275, 6.755, 7.549, 8.183, 9.832, 11.54, 12.029, 12.49, 12.933, 13.29, 13.594, 13.862, 14.09, 14.26, 14.455, 14.588, 14.652, 14.758
-        else
-          gfnc_coeff = 2.679, 3.023, 3.318, 3.664, 3.961, 4.306, 4.651, 4.839, 5.292, 5.636, 5.928, 6.425, 6.841, 8.089, 9.44, 9.875, 10.284, 10.7, 11.05, 11.344, 11.608, 11.831, 12.001, 12.196, 12.329, 12.393, 12.499
-        end
+  def self.gshp_gfnc_coeff(bore_config, g_functions_filename, num_bore_holes, bore_spacing, bore_depth, bore_diameter)
+    require 'json'
+
+    g_functions_filepath = File.join(File.dirname(__FILE__), 'g_functions', g_functions_filename)
```

Comment:
Do we know if these g-functions are "pre-corrected" or not? See https://bigladdersoftware.com/epx/docs/23-1/input-output-reference/group-condenser-equipment.html#field-g-function-reference-ratio. Looks like our current code [assumes a correction has not been applied](https://github.com/NREL/OpenStudio-HPXML/blob/ff35611236bbfb38d3cd026e3f58e63a19c4f625/HPXMLtoOpenStudio/resources/hvac.rb#L287).

Reply:
I'm applying a "correction factor" to these g-functions here: https://github.com/NREL/OpenStudio-HPXML/blob/geothermal_loop/HPXMLtoOpenStudio/resources/hvac_sizing.rb#L2720-L2730.

The correction factor comes from Equation 1 in the g-function library documentation; it's unclear to me whether this is the same "correction factor" referenced by the E+ documentation. If it is, though, it seems we need to enter that value for "G-Function Reference Ratio".

Reply:
Here is the link to download the g-function library zip: https://gdr.openei.org/files/1325/g-function_library_1.0.zip.

In the documentation, scroll to Interpolation Methodology (pdf pg 4). The following refers to "correcting" the g-function values according to:

![image](https://github.com/NREL/OpenStudio-HPXML/assets/8516790/147d36f4-6415-4bf5-9c55-a854c11e663f)

So I think the question is whether the correction term in the previous equation is the same as the one in E+ field "G-Function Reference Ratio". Do we enter the actual radius/length ratio here, thus nullifying any internal calculations?

Reply:
@joseph-robertson I think I've groked my way through the methodology. I've attached the spreadsheet I used to verify calculations. Let me know if something doesn't make sense.

[g-function-library-interp-example.xlsx](https://github.com/NREL/OpenStudio-HPXML/files/12578538/g-function-library-interp-example.xlsx)

1. Select the GHE configuration you're looking to compute. The parameters $r_b$, $H$, $B$, and $D$ can all be different, but the arrangement (e.g. MxN rectangle) should be consistent between all cases. Also, note that the $B/H$ reference values should bound the one you're trying to solve for.
2. Compute the $B/H$ values for the two reference cases and your configuration. This will be used for all interpolations.
3. Interpolate the g-function values using the g-function values from the two reference cases and the $B/H$ values you computed in the previous step.
4. Compute your case's effective $r_b/H$ value by interpolating using the $r_b/H$ values from the reference case and the $B/H$ values from previously. Solve for $r_b$. This will yield an $r_b$ value which is different from the $r_b$ in your specified configuration, which is also sometimes referred to as $r_b^*$ in the docs. 
5. Use the $r_b^*$ (the actual value you want) and the $r_b$ value you computed in the previous step to apply Equation 1 from the docs to "correct" the g-function values. You will see in the docs that the correction is actually quite small for this example.

Hope this helps. Let me know if we need to discuss it further.

Reply:
@mitchute This is a good summary of the methodology, and is what we already have coded up. What I'm looking for is guidance on what should be entered into the "G-Function Reference Ratio" field of the `GroundHeatExchanger:ResponseFactors` object. Previously we had entered a value of 0.0005. Based on the E+ documentation that @shorowit references above, this means that E+ adjusted the g-functions internally. Now that we are following a new methodology for selecting g-function values, should we still enter a value of 0.0005? Or should it be some other value, thus nullifying any internal calculations?

(That said, I tried a few quick runs where I change the reference ratio from very small (e.g., 1e-08) to very large (e.g., 1), and it appears to have no affect on results. 🤷‍♂️)

Reply:
Got it. As I think I mentioned when we discussed it, the current implementation is clunky (my fault),  so you just need to ensure that the ratio ($r_b/H$) is consistent with your $r_b$ and $H$ values. If you do that, the g values will be taken as-is. Otherwise, it will attempt to reapply the correction noted in Equation 1.

https://github.com/NREL/EnergyPlus/blob/2ad5b542c761abde2179225b08ce7af0b94a5d3b/src/EnergyPlus/GroundHeatExchangers.cc#L2882-L2900

Reply:
@mitchute Below you can see the value I've entered for "G-Function Reference Ratio". It is `(0.127 / 2.0) / 83.5152 = 0.000760340632603406` (i.e., equal to `RATIO` in your link). So I would therefore expect different results than if the field was set to `0.0005`, but the results are the same. Perhaps I'm running a case where the correction is too small to make any difference?

```
GroundHeatExchanger:Vertical:Properties,
  ground source heat pump exchanger Properties, !- Name
  1,                                      !- Depth of Top of Borehole {m}
  83.5152,                                !- Borehole Length {m}
  0.127,                                  !- Borehole Diameter {m}
  0.6924,                                 !- Grout Thermal Conductivity {W/m-K}
  3900000,                                !- Grout Thermal Heat Capacity {J/m3-K}
  0.39813,                                !- Pipe Thermal Conductivity {W/m-K}
  1770000,                                !- Pipe Thermal Heat Capacity {J/m3-K}
  0.02667,                                !- Pipe Outer Diameter {m}
  0.0024257,                              !- Pipe Thickness {m}
  0.05120894;                             !- U-Tube Distance {m}

Site:GroundTemperature:Undisturbed:KusudaAchenbach,
  ground source heat pump exchanger Ground Temps, !- Name
  1.731,                                  !- Soil Thermal Conductivity {W/m-K}
  920,                                    !- Soil Density {kg/m3}
  3506.47993311037,                       !- Soil Specific Heat {J/kg-K}
  10.8753424657534,                       !- Average Soil Surface Temperature {C}
  3.2,                                    !- Average Amplitude of Surface Temperature {deltaC}
  8;                                      !- Phase Shift of Minimum Surface Temperature {days}

GroundHeatExchanger:ResponseFactors,
  ground source heat pump exchanger Response Factors, !- Name
  ground source heat pump exchanger Properties, !- GHE:Vertical:Properties Object Name
  3,                                      !- Number of Boreholes
  0.000760340632603406,                   !- G-Function Reference Ratio {dimensionless}
```

Reply:
@joseph-robertson it looks like the ratio input into the `ResponseFactors` object is overridden here, https://github.com/NREL/EnergyPlus/blob/2ad5b542c761abde2179225b08ce7af0b94a5d3b/src/EnergyPlus/GroundHeatExchangers.cc#L383-L383 

so, it's not even used... 😬

My suggestion is to make sure the correction is pre-applied to the g-functions so they are consistent with BH radius and height. I'll submit a PR to fix this so the ratio input value can be used.

Reply:
@joseph-robertson Has this been addressed? I see that the conversation has not been resolved.

Reply:
Yes, I believe so. We have:
```
ground_heat_exch_vert.setGFunctionReferenceRatio(ground_heat_exch_vert.boreHoleRadius.get / ground_heat_exch_vert.boreHoleLength.get) # ensure this ratio is consistent with rb/H so that g values will be taken as-is
```
This ensures that no further correction is applied to the g-function values.